### ISSUE TITLE
Npm install is too expensive

### ISSUE THREAD

**Issue by vallettea** [anthill/6brain#68]
Yesterday we update 6bin on a sensor and we accidentally did it over 3G. The reinstall of 6bin twice consumed 10Mo of data.
We need to optimise this:
- use of npm 3+ for flat dependencies
- does npm compresses the modules before sending them over the air ?
- should we use git to send only the diffs we need ?

**Comment by DavidBruant** [comment]
For more context on this issue:
- Data sent over 3G on this particular project costs us money (~1€/MB)
- Docker or other containerization technology is not an option in the short term (because of the OS we chose on Raspberry PI, a choice we made because of some other contraints)

**Comment by vallettea** [comment]
@DavidBruant i managed to install node 4.2.3 and npm 3.5 on rpi !

**Comment by vallettea** [comment]
@oncletom I did not get your remark on twitter: _you should not perform an 'npm up': your CI generate an artefact of successful build, you deploy a diff/zip_

**Comment by vallettea** [comment]
A `npm install` with no new package installed cost 300ko up and 300ko down (for this package).

**Comment by vallettea** [comment]
installing 6bin costs 2.6Mo

**Comment by thom4parisot** [comment]
Will do – did not manage to find the time today.

tl;dr pushing a tar file (or a diff of two tarballs) of an install which has proven to work is more reliable than running as many installs/updates as devices (which means network roundtrips over 3G for nothing and race conditions/bugs inherent to npm/network which make the install state unpredictable).

Similar to Docker: you push an image you ensured the install + isolated tests were successful instead of running the install as many times as you deploy.

You just need to build on the same hardware because of you have native modules but you can automate the build on one of your device or use a Raspberry Pi cloud provider). Or you can also ignore deploying binaries and call `npm rebuild` on the devices, should be even lighter to deploy.

**Comment by vallettea** [comment]
Thanks @oncletom 
Never heard of `npm rebuild` before. I guess you're right, we should try to mimic docker. 
We stopped using the hypriot image that enables to use docker on rpi for various reasons but we could think again. 
So here are the alternatives I see:

### Docker
- [] check we can install all we need on the hypriot distribution
- [] check our reverse ssh and ansible capability is compatible with the container (networking through containers is not always easy)
- [] see the load on the wire: when you modify your code, you build the new image locally and push it to the sensors, how much data is consumed ? Is the layered content of the image properly handled ?

### Npm
- [] check about rebuild (doesn't it uses the network at all)
- [] combine ansible to create an compressed archive that is sent on the wire and then installed through npm rebuild

### Git
- [] read about the efficiency of github compression
- [] how to install the dependencies ?

**Comment by vallettea** [comment]
interesting for the docker solution: 
https://resin.io/blog/building-arm-containers-on-any-x86-machine-even-dockerhub/
